File: src/main/java/src/com/study/design/pattern/solid/single/responsibility/Journal.java
package src.com.study.design.pattern.solid.single.responsibility;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Handles the responsibility for add and remove entries in journal
 */
class Journal {

    private final List<String> entries = new ArrayList<>();

    void addEntry(String text) {
        entries.add(getIndex() + ": " + text);
    }

    void removeEntry() {
        if (entries.isEmpty())
            return;

        entries.remove(entries.size() - 1);
    }

    private Integer getIndex() {
        return entries.size() + 1;
    }

    @Override
    public String toString() {
        return String.join(System.lineSeparator(), entries);
    }

    public void save(String filename) throws Exception {

        System.out.println("This method violate the single responsibility principle by adding a new concern to Journal: Persistence. Shouldn't be here!");

        try (PrintStream out = new PrintStream(filename)) {
            out.println(toString());
        }
    }
}

File: src/main/kotlin/src/com/study/design/pattern/solid/single/responsibility/Journal.kt
package src.com.study.design.pattern.solid.single.responsibility

import java.io.PrintStream

/**
 * Handles the responsibility for add and remove entries in journal
 */
class Journal {

    val entries = mutableListOf<String>()

    fun addEntry(text: String) = entries.add("${entries.size} : $text")

    fun removeEntry() {
        if (entries.isNotEmpty()){
            entries.removeAt(entries.lastIndex)
        }
    }

    private fun save(filename: String) {
        print("This method violate the single responsibility principle by adding a new concern to Journal: Persistence. Shouldn't be here!")
        PrintStream(filename).println(toString())
    }

    override fun toString() = entries.joinToString(System.lineSeparator())
}